contrib/babelfishpg_tsql/src/tsqlIface.cpp

```diff
@@ -1666,9 +1671,86 @@ class tsqlBuilder : public tsqlCommonMutator
 		}
 
+		if (ctx->execute_statement()){
+			PLtsql_stmt_exec *stmt = (PLtsql_stmt_exec *) getPLtsql_fragment(ctx);
+			if (stmt->cmd_type == PLTSQL_STMT_EXEC && stmt->proc_name && pg_strcasecmp("sp_tables", stmt->proc_name) == 0){
+				PLtsql_expr_query_mutator mutator(stmt->expr, ctx);
+				add_rewritten_query_fragment_to_mutator(&mutator); // move information of rewritten_query_fragment to mutator.
+				mutator.run(); // expr->query will be rewitten here
+			}
+		}
+
 		clear_rewritten_query_fragment();
 	}
 
+	// Try to retrieve the procedure name as in execute_body/execute_body_batch
+	// from the context of execute_parameter.
+	static std::string getProcNameFromExecParam(TSqlParser::Execute_parameterContext *exParamCtx)
+	{
+		antlr4::tree::ParseTree *ctx = exParamCtx;
+		// We only need to loop max number of dereferences needed from the context 
+		// of execute_parameter to execute_body/execute_body_batch. 
+		// Currently that number is 3 + (number of execute_statement_arg_unnamed - 1).
+		// Because for now we only need this function for sp_tables which have 
+		// at most 5 parameters, 8 seems to be good enough here.
+		int cnt = 8; 
+
+		while (ctx->parent != nullptr && cnt-- > 0)
+		{
+			TSqlParser::Execute_bodyContext * exBodyCtx = dynamic_cast<TSqlParser::Execute_bodyContext *>(ctx->parent);
+			TSqlParser::Execute_body_batchContext *exBodyBatchCtx = dynamic_cast<TSqlParser::Execute_body_batchContext *>(ctx->parent);
+			TSqlParser::IdContext *proc = nullptr;
+
+			if (exBodyCtx != nullptr)
+				proc = exBodyCtx->func_proc_name_server_database_schema()->procedure;
+			else if (exBodyBatchCtx != nullptr)
+				proc = exBodyBatchCtx->func_proc_name_server_database_schema()->procedure;
+
+			if (proc != nullptr)
+				return stripQuoteFromId(proc);
+
+			ctx = ctx->parent;
+		}
+		return "";
+	}
+
+	void exitExecute_parameter(TSqlParser::Execute_parameterContext *ctx) override
+	{
+		std::string argStr;
+		if (ctx->constant())
+			argStr = getFullText(ctx->constant());
+		else if (ctx->id())
+			argStr = getFullText(ctx->id());
+
+		if (ctx->constant() || ctx->id())
+		{
+			// BABEL-2395, BABEL-3640: embedded single quotes are allowed in procedure parameters.
+			// in enterExecute_parameter, we replace "" with '', and also record the place of "" into 
+			// double_quota_places, then in here, we'll replace all ' into '' if it's inside ""
+			if (pg_strcasecmp(getProcNameFromExecParam(ctx).c_str(), "sp_tables") == 0)
+			{
+				std::string newStr;
+				if (argStr.size() > 1 && argStr.front() == '\'' && argStr.back() == '\''
+				&& std::binary_search (double_quota_places.begin(), double_quota_places.end(), parameterIndex)){
+					newStr += '\'';
+					for (std::string::iterator iter = argStr.begin() + 1; iter != argStr.end() - 1; ++iter)
+					if (*iter == '\''){
+						newStr += "\'\'";
+					} else{
+						newStr += *iter;
+					}
+					newStr += '\'';
+					if (ctx->constant())
```

Comment:
It seems like `newStr` and the above for-loop are only needed if `ctx->constant()` is true. Let's make the code more efficient then.

Reply:
The code shows it also possible ctx->id() is true